Add support for `-sEXPORT_ES6`/`*.mjs` on Node.js

kleisauke opened the pull request:
As described in #11792, `require()` and `__dirname` doesn't exist in
an ES6 module. Emscripten uses this to import built-in core Node.js
modules. For example, the `node:fs` module is used for synchronously
importing the `*.wasm` binary, when not linking with `-sSINGLE_FILE`.

To work around this, ES6 modules on Node.js may import
`createRequire()` from `node:module` to construct the `require()`
function, allowing modules to be imported in a CommonJS manner.

Emscripten targets a variety of environments, which can be
categorized as:
1. Multi-environment builds, which is the default when
   `-sENVIRONMENT=*` is not specified at link time.
2. Single-environment, e.g. only web or Node.js as target.

For use case (1), this commit ensures that an `async` function is
emitted, allowing Node.js modules to be dynamically imported. This is
necessary given that static import declarations cannot be put in
conditionals. Inside the module, for Node.js only, it's using the
above-mentioned `createRequire()`-construction.

For use case (2), when only Node.js is targeted, a static import
declaration utilize the same `createRequire()`-construction.

For both use cases, `-sUSE_ES6_IMPORT_META=0` is not allowed, when
Node.js is among the targets, since it is not possible to mimic
`__dirname` when `import.meta` support is not available.

This commit does not change anything for use case (2), when only the
web is targeted (`-sENVIRONMENT=web`).

Resolves: #11792.
Resolves: #17797.

sbc100 commented on src/preamble.js:
```diff
@@ -649,7 +649,8 @@ if (Module['locateFile']) {
   // Use bundler-friendly `new URL(..., import.meta.url)` pattern; works in browsers too.
-  wasmBinaryFile = new URL('{{{ WASM_BINARY_FILE }}}', import.meta.url).toString();
+  // FIXME: On Node.js this needs to be wrapped in fileURLToPath.
+  wasmBinaryFile = new URL('{{{ WASM_BINARY_FILE }}}', import.meta.url).href;
```
That the `toString` -> `href` change?

sbc100 replied:
Nice!

kleisauke commented:
This should be ready for review, but I'm unsure how to proceed due to this `FIXME`:
https://github.com/emscripten-core/emscripten/blob/bc619d3fbc897f8d1349dfe1fe17266d0ccbcf75/src/shell.js#L205-L208

Also, perhaps we should swap all `require()`'s with `import()`'s?
https://github.com/emscripten-core/emscripten/blob/bc619d3fbc897f8d1349dfe1fe17266d0ccbcf75/emcc.py#L3596

But AFAIK that requires a lot more changes, and I'm not sure it's worth it. For example, should we then eagerly import the crypto module too?
https://github.com/emscripten-core/emscripten/blob/f9d26b08403737dcb7aa8c249057908a89e28b7a/src/library.js#L2229-L2238

/cc @curiousdannii the creator of issue #11792.

sbc100 commented on emcc.py:
```diff
@@ -3016,6 +3016,7 @@ def phase_final_emitting(options, state, target, wasm_target, memfile):
     src = read_file(final_js)
     final_js += '.esmeta.js'
     write_file(final_js, src.replace('EMSCRIPTEN$IMPORT$META', 'import.meta'))
+    shared.get_temp_files().note(final_js)
```
What this temp file leaking previously?

kleisauke replied:
Yes, this was noted in this PR using the `EMTEST_DETECT_TEMPFILE_LEAKS=1` env variable, which is set by default in CI. I could make a separate PR for this, if you like.

sbc100 commented on src/closure-externs/node-externs.js:
```diff
@@ -111,3 +111,13 @@ Buffer.prototype.slice = function(start, end) {};
+/**
+ * @param {URL|string} url
+ */
+var fileURLToPath = function(url) {};
```
According to the docs seem to exist on the `url` module, rather than at the top level..?

brendandahl replied:
They are, but they're imported into the module's global scope (see `emcc.py`).

sbc100 replied:
But shouldn't closure be able to see that then?  It seems odd to me that we should need to declare these in this global location like this?  Perhaps a local annotation where they are declared/imported?

brendandahl replied:
It does seem odd. Maybe closure doesn't understand `await import('url')`?

kleisauke replied:
Indeed, it's a bit odd that this is needed. I think this was necessary when targeting both web/node as well as node only. That is:
```js
if (typeof process == 'object' &&
    typeof process.versions == 'object' &&
    typeof process.versions.node == 'string') {
  const { createRequire } = await import('module');
  var { pathToFileURL, fileURLToPath } = await import('url');
  var require = createRequire(import.meta.url);
}
```
(`-sENVIRONMENT=web,node`)
```js
import { createRequire } from 'module';
import { pathToFileURL, fileURLToPath } from 'url';
const require = createRequire(import.meta.url);
```
(`-sENVIRONMENT=node`)

Perhaps closure is run on an indeterminate file instead of the final JS file with these imports? Let me verify this.

kleisauke replied:
It seems that Closure indeed has some limitations with dynamic import expressions.
https://github.com/google/closure-compiler/wiki/JS-Modules#dynamic-import-expressions

I moved these externals to the correct file with commit 68377b6069d5bd3fad45965a66ee40bed9cecc46 and removed the need of these `import()`'s with commit bae5bf5bd5b7792ae93a085f96fc8a8637f52876.

sbc100 commented on src/shell.js:
```diff
@@ -193,11 +193,25 @@ if (ENVIRONMENT_IS_NODE) {
+  // These modules will usually be used on Node.js. Load them eagerly to avoid
+  // the complexity of lazy-loading.
+  var fs = require('fs');
+  var nodePath = require('path');
```
In moving these require statements we've also lost that long comment explaining them.  Was the deliberate?

kleisauke replied:
That was intentional. That comment mentioned that it was building an ES6 module where `require()` did not exist. This (draft) PR would solve that issue.

We could argue for excluding these imports when building with `-sSINGLE_FILE`. But in that case, `stdin` usage, or using the `-sMAIN_MODULE` and/or `-sNODERAWFS` options would still depend on these imports (as noted in PR https://github.com/emscripten-core/emscripten/pull/17851).

sbc100 commented on test/test_other.py:
```diff
@@ -237,41 +237,57 @@ def test_emcc_generate_config(self):
     '': (True, [],),
     'no_import_meta': (False, ['-sUSE_ES6_IMPORT_META=0'],),
   })
+  @node_pthreads
```
Why pthreads here?

kleisauke replied:
The test links with `-pthread`, so Node.js before version 16 needs the `--experimental-wasm-bulk-memory --experimental-wasm-threads` flags to successfully execute the `subdir/hello_world.mjs` script.

sbc100 reviewed (approved):
lgtm with some comments

curiousdannii commented:
This is a clever temporary solution, but I'm pretty sure it wouldn't help when the code is bundled, so it's not a full solution to #11792. Substantial parts of the JS output will need to be redesigned in order to be async, so as I've said before, before #11792 can be properly solved it will need direction from the core Emscripten team.

sbc100 commented:
> This is a clever temporary solution, but I'm pretty sure it wouldn't help when the code is bundled, so it's not a full solution to #11792. Substantial parts of the JS output will need to be redesigned in order to be async, so as I've said before, before #11792 can be properly solved it will need direction from the core Emscripten team.

Are you ok with landing this as a temporary fix?  It seems like have some test coverage is better than none.

curiousdannii commented:
I haven't tested it myself, but the idea is a good one, and it would definitely be good to get some more test coverage.

brendandahl commented:
> cc @brendandahl who might know this ES6 stuff? I'm afraid I don't know much about it myself...

Looks and works how'd I'd expect.

kleisauke commented:
I resolved the above mentioned `FIXME` by ignoring `-sUSE_ES6_IMPORT_META=0` when targeting Node, see commit bae5bf5bd5b7792ae93a085f96fc8a8637f52876. AFAIK, this option was introduced to fix compat on old browsers and bundlers.

RReverser commented:
> Is there a design document of how this would look like? I suppose Emscripten needs to change this emit:

Emscripten already has a mechanism for async dependencies, check out `addRunDependency` and the surrounding code. That's what takes care of e.g. Wasm async compilation, and some other things that need to happen before startup. And then, there is already a promise in `Module.ready`, so adding yet another one at the top level (due to async function) would be inconsistent.

RReverser commented:
> I suppose Emscripten needs to change this emit:

Actually... maybe best to do that as part of this PR? Then it wouldn't depend on top-level await, which is a relatively new feature in browsers (definitely newer than `import.meta` and we use conditional checks even for that) and still not supported by some tooling/bundlers.

kleisauke commented:
> Then it wouldn't depend on top-level await, which is a relatively new feature in browsers (definitely newer than `import.meta` and we use conditional checks even for that)

Ah, I think you're right. Let me fix that. I assumed that guarding the top-level await import for Node.js would prevent browsers from failing, but this probably didn't work for Node.js before v14.8 either.

kleisauke commented:
... should be fixed with commit e8f83e986b5543bfe35c98cf3633cdc3d3d1e719.

RReverser commented on src/preamble.js:
```diff
@@ -653,7 +653,13 @@ if (Module['locateFile']) {
+  wasmBinaryFile = new URL('{{{ WASM_BINARY_FILE }}}', import.meta.url);
+#if ENVIRONMENT_MAY_BE_NODE
+  if (ENVIRONMENT_IS_NODE) {
+    wasmBinaryFile = nodeUrl.fileURLToPath(wasmBinaryFile);
```
I think this step shouldn't be necessary.

Node.js `fs.readFile` can read files by their `file://` URLs in all versions that have ESM support (in fact, construction of URLs relative to `import.meta.url` was one of the main reasons for adding URL support to `fs`).

kleisauke replied:
Good point! Should be fixed with commit cff5945db24ffc560421f28759efc2993f931925, I only needed to convert the `file://` string back into URL to allow it to pass to `fs.readFileSync`, but that's all.

RReverser commented on src/shell.js:
```diff
@@ -193,11 +193,35 @@ if (ENVIRONMENT_IS_NODE) {
+    // EXPORT_ES6 + ENVIRONMENT_IS_NODE always requires use of import.meta.url,
+    // since there's no way getting the current absolute path of the module when
+    // support for that is not available.
+    scriptDirectory = nodePath.dirname(nodeUrl.fileURLToPath(import.meta.url));
```
This one I'm less sure about, but also worth checking if we can get away with just keeping the URL here.

kleisauke replied:
Indeed, I wasn't sure about this. I removed the `fileURLToPath` construction with commit  cff5945db24ffc560421f28759efc2993f931925, however, that may cause the `locateFile` function to return `file://` strings on Node.js.
https://github.com/emscripten-core/emscripten/blob/84a634167a1cd9e8c47d37a559688153a4ceace6/src/shell.js#L152-L161
Let's see if the CI likes that, though locally it still passes the `other.test_emcc_output_mjs* other.test_export_es6*` tests.

RReverser replied:
> however, that may cause the `locateFile` function to return `file://` strings on Node.js.

Now I'm really curious, if `file://` strings are not accepted directly, then how would same strings returned from `locateFile` work...

kleisauke replied:
> Now I'm really curious, if `file://` strings are not accepted directly, then how would same strings returned from `locateFile` work...

IIUC, `file://` strings are accepted after that `isFileURI(filename) ? new URL(filename) ...` change in `read_`, which will probably always be called on Node.js when reading files. But maybe it lacks some test coverage for this.

RReverser replied:
Oh ok, that makes sense. (well, at least if nothing else uses the result of `locateFile`)

kleisauke replied:
I suppose it could also cause issues if someone does this:
```js
Module['locateFile'] = function (fileName, scriptDirectory) {
  if (scriptDirectory.startsWith('file://')) {
    throw new TypeError();
  }
  return scriptDirectory + fileName;
}
```
(i.e. overriding `Mode.locateFile` with custom logic based on `scriptDirectory`)

But I'm not sure if this is done in the wild.

RReverser replied:
Maybe it's safer to keep scriptDirectory as actual filepath then, at least for now.

kleisauke replied:
9074bb71ac76b35057cb27ca03be5f864a3f9444 :)

kleisauke replied:
Though, while doing that, I noticed this:
https://github.com/emscripten-core/emscripten/blob/84a634167a1cd9e8c47d37a559688153a4ceace6/src/shell.js#L136-L137
... a couple of lines further ...
https://github.com/emscripten-core/emscripten/blob/84a634167a1cd9e8c47d37a559688153a4ceace6/src/shell.js#L364-L370
So, perhaps `scriptDirectory` was already a `file://`-string on Node.js when linking with `-sEXPORT_ES6`? Let me verify this.

kleisauke replied:
... nvm, I didn't look carefully, that is guarded with `#if SHARED_MEMORY && !MODULARIZE`, though `MODULARIZE` is forcefully enabled when linking with `-sEXPORT_ES6`:
https://github.com/emscripten-core/emscripten/blob/89a494fa1b1c6e10210d8e1e87dd9f475dd6b4a1/emcc.py#L2253-L2257
So I'm not sure if this comment is up-to-date:
https://github.com/emscripten-core/emscripten/blob/84a634167a1cd9e8c47d37a559688153a4ceace6/src/shell.js#L134-L135

RReverser commented on src/node_shell_read.js:
```diff
@@ -11,7 +11,10 @@ read_ = (filename, binary) => {
   }
 #endif
-  filename = nodePath['normalize'](filename);
+  // We need to re-wrap `file://` strings to URLs. Normalizing isn't
```
Wait, why? The string is already a URL, you don't need to create an `URL` object (which Node.js will convert once again to string anyway).

RReverser replied:
I think you can actually remove this line altogether - normalize (that was here before your change) shouldn't be necessary either, as `readFileSync` happily accepts non-normalized / relative URLs too.

RReverser replied:
So `filename`, whatever it is - whether a URL string or file path - can just be passed unmodified to `readFileSync`.

kleisauke replied:
IIUC, the `.href` property (formerly the `.toString()` call) when the `wasmBinaryFile` is assigned made this a `string`.

RReverser replied:
Huh :/ I was sure it accepts file URLs as strings and that I used it as such... Sorry for misleading.

RReverser commented on src/shell.js:
```diff
@@ -193,11 +193,34 @@ if (ENVIRONMENT_IS_NODE) {
+    // EXPORT_ES6 + ENVIRONMENT_IS_NODE always requires use of import.meta.url,
+    // since there's no way getting the current absolute path of the module when
+    // support for that is not available.
+    scriptDirectory = nodePath.dirname(import.meta.url);
```
`nodePath` might not always work correctly on URLs. I'd rather use `URL` itself for url manipulation, so here it'd be `new URL('./', import.meta.url).href` (which also ensures trailing slash).

kleisauke replied:
Ah, nice! Let me incoperate that.

RReverser replied:
We should probably update the web version of this at some point too. https://github.com/kleisauke/emscripten/blob/cff5945db24ffc560421f28759efc2993f931925/src/shell.js#L401

Although maybe it was written that way because `URL` class didn't exist in some older browsers yet... Just something to look into in the future.

RReverser replied:
I mean, in the end, ideally `scriptDirectory` should be always inferred from `import.meta.url` on any `EXPORT_ES6` + (use meta) target. That's what I tried to do first in my attempt at tackling that issue and failed because of many conflicting codepaths 😅

But we can leave that out of scope.

kleisauke replied:
Indeed, `_scriptDir` and `scriptDirectory` needs some careful refactoring. I didn't investigate why this was separated into two vars initially, but perhaps I should.

For now, I incorporated your first suggestion with commit 85a9619c414757a6cfbd784fd8a57541527d30d0.

kleisauke commented on src/shell.js:
```diff
@@ -193,11 +193,34 @@ if (ENVIRONMENT_IS_NODE) {
+
   if (ENVIRONMENT_IS_WORKER) {
-    scriptDirectory = require('path').dirname(scriptDirectory) + '/';
+    scriptDirectory = nodePath.dirname(scriptDirectory);
```
IIUC, it's a string path. But perhaps this should be?:
```suggestion
    scriptDirectory = nodePath.dirname(_scriptDir) + '/';
```
See e.g. commit https://github.com/emscripten-core/emscripten/pull/10343/commits/5decc7028f22142b87cec134ba6e39b976d22094 where this was introduced.

RReverser commented:
@kleisauke Do you want to unmark this as "draft" PR? I think it's pretty close to being mergeable. Assuming the tests pass after 6e5201640984e1cc56d671c5f8c2e11a03005ff6 at least :)

kleisauke commented:
This is ready for review now, diff is best viewed with the "Hide whitespace" option.

kleisauke commented:
... commit message still needs some polishing. It should at least contain these details when `EXPORT_ES6` is used in combination with:
- Multi-environment builds (`ENVIRONMENT=web,node`) - uses `await import` and emits an async function;
- Only targeting Node.js (`ENVIRONMENT=node`) - uses a static import declaration;
- Only targeting web (`ENVIRONMENT=web`) - no changes;
- Targeting node (`ENVIRONMENT=*node*`) and using `USE_ES6_IMPORT_META=0` is disallowed (i.e. `import.meta.url` is always emitted).

RReverser commented:
> It should at least contain these details when `EXPORT_ES6` is used in combination with:

Heh, ideally EXPORT_ES6 with those things you mentioned should be documented in actual docs, not (just) commit message. See #14383 if you want to tackle that too :)

kleisauke commented on test/test_other.py:
```diff
@@ -237,41 +237,63 @@ def test_emcc_generate_config(self):
-    '': (True, [],),
-    'no_import_meta': (False, ['-sUSE_ES6_IMPORT_META=0'],),
+    '': ([],),
+    'node': (['-sENVIRONMENT=node'],),
```
> that doesn't allow it to be run with Node.js

... forgot to mention that the `test_other.py` suite is only executed with Node.js or V8, as far as I know.

RReverser reviewed (approved):
I think this is good to go. We should still refactor the whole `_scriptDir` / `scriptDirectory` entanglement at some point, but as it is, this should allow ESM to be finally used on Node.js target, so it's already a great improvement.

RReverser commented:
Ah well, I approved and only then noticed the CI failure 😅 Not sure if due to tests or some spurious issue - @kleisauke can you please take a look?

kleisauke commented:
Commit 387dba83b13fde2a603908e714075a6baf8e025f only changed `test_other.py`, so it should _not_ affect the `test-browser-chrome` (`test_browser.py`) suite. It's probably a flake.

Is the commit history still relevant for reviewing purposes? If not, I can squash, polish, rebase and (force-)push it.

RReverser commented:
> Is the commit history still relevant for reviewing purposes? If not, I can squash, polish, rebase and (force-)push it.

Personally, I don't think so, but let's see what @sbc100 says as he's also been reviewing.

sbc100 commented:
> > Is the commit history still relevant for reviewing purposes? If not, I can squash, polish, rebase and (force-)push it.
> 
> Personally, I don't think so, but let's see what @sbc100 says as he's also been reviewing.

Please go ahead and squash whenever you please.    There is no requirement to do this before landing (but it is required as part of the final commit).

kleisauke commented:
Great! I just squashed, rebased, and polished the commit message. I also opened PR https://github.com/WebAssembly/binaryen/pull/5235, since I think that `-sSINGLE_FILE` workaround could cause issues with this line:
https://github.com/emscripten-core/emscripten/blob/b510fdd7ad8d21c965bbf0591f77c4cd41480d36/src/shell.js#L213
(i.e., it ends up calling `{}.fileURLToPath(new URL('./', import.meta.url))`)

kripken reviewed (commented):
I don't understand the JS side here fully, but very happy to see the progress here! And overall things lgtm when I do understand them :smile: 

1. Please add a changelog entry.
2. Do we have an existing test for running this output in a browser?
3. Should we wait on https://github.com/WebAssembly/binaryen/pull/5235 before landing?

RReverser commented:
> (i.e., it ends up calling `{}.fileURLToPath(new URL('./', import.meta.url))`)

Hm, why if you're redefining `require` to a different value? Wouldn't [binaryen.js-extern-pre.js](https://github.com/WebAssembly/binaryen/pull/5235/files#diff-6b2403b57b48312f7abce5c75138a56294b70f1066def96b83f7c54751fc2fa7) be executed in preamble before your redefinition of `require` anyway?

lgarron reviewed (commented):
Is there an easy way to test out this change? I'm eager to test WASM in ESM in `node`!

kleisauke commented:
I was being to cautious, it looks like binaryen should not be affected, since it does a multi-environment build. Such workaround would only causes issues if you link with `-sENVIRONMENT=node`.

For example:
```js
import { createRequire } from 'module';
const require = createRequire(import.meta.url);
...
function require() {
  return {};
}
...
var fs = require('fs');
var nodePath = require('path');
...
scriptDirectory = require('url').fileURLToPath(new URL('./', import.meta.url)); // includes trailing slash
```

```
scriptDirectory = require('url').fileURLToPath(new URL('./', import.meta.url)); // includes trailing slash
                                 ^
TypeError: require(...).fileURLToPath is not a function
```

@kripken

1. I'll add the missing changelog entry shortly, thanks for notifying!
2. There's currently at least two browser tests that uses `-sEXPORT_ES6`, but perhaps we should increase that test coverage.
3. Given the above, I don't think https://github.com/WebAssembly/binaryen/pull/5235 should block this PR and may land independently without breaking CI.

@lgarron I don't know if there is an easy way to test this. I vaguely remembered that there was an emsdk setting that could easily swap the Emscripten repository to point to a fork instead, but I'm not sure which setting that was.

@RReverser WDYT about commit 52444ee16bd6a62d8fe92668d65f326006d19797? It should probably be split into its own PR, but that would allow Closure to optimize the wrapper as well. For example:
```js
var Module=(()=>{var P=import.meta.url;return async function(b){<CODE>;return b.ready}})();
export default Module;
```

kleisauke commented:
@lgarron I found it, it was the `--override-repository` command introduced in commit https://github.com/emscripten-core/emsdk/commit/3510ab2697eff7317d449c7beee2b5acd607304f.

So, to test this PR with the tip-of-tree compilers, you could use:
```bash
$ cd emsdk/
$ git pull
$ ./emsdk install tot emscripten-main-64bit --override-repository emscripten-main-64bit@https://github.com/kleisauke/emscripten/tree/test-es6-node
$ ./emsdk activate tot emscripten-main-64bit
$ source ./emsdk_env.sh

# Verify commit
$ git -C emscripten/main rev-parse HEAD
52444ee16bd6a62d8fe92668d65f326006d19797
# Verify emcc
$ which emcc
~/emsdk/emscripten/main/emcc

# Invalidate cache, if any 
$ emcc --clear-cache
```

sbc100 commented:
> @lgarron I found it, it was the `--override-repository` command introduced in commit [emscripten-core/emsdk@3510ab2](https://github.com/emscripten-core/emsdk/commit/3510ab2697eff7317d449c7beee2b5acd607304f).
> 
> So, to test this PR with the tip-of-tree compilers, you could use:
> 
> ```shell
> $ cd emsdk/
> $ git pull
> $ ./emsdk install tot emscripten-main-64bit --override-repository emscripten-main-64bit@https://github.com/kleisauke/emscripten/tree/test-es6-node
> $ ./emsdk activate tot emscripten-main-64bit
> $ source ./emsdk_env.sh
> 
> # Verify commit
> $ git -C emscripten/main rev-parse HEAD
> 52444ee16bd6a62d8fe92668d65f326006d19797
> # Verify emcc
> $ which emcc
> ~/emsdk/emscripten/main/emcc
> 
> # Invalidate cache, if any 
> $ emcc --clear-cache
> ```

You don't even need to go that far.   You can just install and activate emsdk, but then put your emscripten checkout first in the PATH after activation.

i.e. when you run `./emcc` from your checkout it will always use you local version of emscripten.

kleisauke commented:
> You don't even need to go that far. You can just install and activate emsdk, but then put your emscripten checkout first in the PATH after activation.
> 
> i.e. when you run `./emcc` from your checkout it will always use you local version of emscripten.

You're right. But that requires a few more steps and this automatically runs `npm ci` and installs the platform-specific Closure Compiler for you.

Also, you can install and activate `emscripten-main-64bit` independently, which might be convenient on GitHub Actions. For example, see commit https://github.com/WebAssembly/binaryen/pull/5235/commits/e858ca18625d49af112b31840dc358a8981ef687.

kleisauke commented:
I've dropped commit 234940f032e4353a4390af02bbc59a6a9f87afcf, ee3c31a0985b47fcd3df78aa16a93cfa6757f8c1, 133f11c6dde60173df158455882ef1ee2b8201b7 and 7a377d264f82c80f049712e899fbc4e262eedfd3 for now, since that's a larger change than initially addressed in commit 79c2fef3afaa2e35ac7b708f4d157e105fe528a5. I'll make a separate PR for those changes after this PR lands.

kripken commented:
Ok, great, sounds good to me to land then! Does anyone else have any remaining comments?

I think it would be good to add some docs for this under `site/` - not sure where, perhaps a new page for ES6 support specifically?  Or perhaps it can fit in the "interacting with JS" page. But that can be separate from this PR.

sbc100 commented on emcc.py:
```diff
@@ -3582,6 +3583,35 @@ def preprocess_wasm2js_script():
+  else:
+    return '''
+import { createRequire } from 'module';
+import { pathToFileURL, fileURLToPath } from 'url';
```
Sorry I think this might have been an old comment that I hadn't send.. maybe can ignore this?

kleisauke replied:
No worries. The issue was that this modularize wrapper and imports are appended after Closure in ran, so it's not unaware of this. For the same reason, it currently requires the `createRequire` definition in `closure-externs.js` to avoid minification, if any.
https://github.com/emscripten-core/emscripten/blob/5ec36415577b9f8492c843134dc6e38ec48f4438/src/closure-externs/closure-externs.js#L18-L19

I avoided the use of the `pathToFileURL` and `fileURLToPath` imports here, since it's now only used by:
https://github.com/emscripten-core/emscripten/blob/5ec36415577b9f8492c843134dc6e38ec48f4438/src/shell.js#L213

For reference, here are the new static import declarations:
https://github.com/emscripten-core/emscripten/blob/5ec36415577b9f8492c843134dc6e38ec48f4438/emcc.py#L3702-L3706

kleisauke replied:
... I also tried to move these static imports to `pre_modularize` in commit 234940f032e4353a4390af02bbc59a6a9f87afcf locally, but that seems to cause a `java.lang.NullPointerException` when running `other.test_emcc_output_mjs_closure_node`.
<details>
  <summary>Details</summary>

```
building:ERROR: Closure compiler run failed:

building:ERROR: java.lang.NullPointerException: NAME createRequire$$module$module 3:16  [length: 13] [source_file: /tmp/tmpvxnkv4rn/emscripten_temp_012it4fl/hello_world.js.pre-modular.js] [original_name: createRequire] : Color{id=0, debugInfo=DebugInfo{compositeTypename=unknown}, prototypes=[], instanceColors=[], invalidating=true, propertiesKeepOriginalName=false, constructor=false, ownProperties=[], boxId=null, closureAssert=false, unionElements=[]}
	at com.google.common.base.Preconditions.checkNotNull(Preconditions.java:907)
	at com.google.javascript.jscomp.RemoveUnusedCode.getVarForNameNode(RemoveUnusedCode.java:841)
	at com.google.javascript.jscomp.RemoveUnusedCode.traverseNameNode(RemoveUnusedCode.java:694)
	at com.google.javascript.jscomp.RemoveUnusedCode.traverseNode(RemoveUnusedCode.java:520)
	at com.google.javascript.jscomp.RemoveUnusedCode.traverseChildren(RemoveUnusedCode.java:1245)
	at com.google.javascript.jscomp.RemoveUnusedCode.traverseCall(RemoveUnusedCode.java:752)
	at com.google.javascript.jscomp.RemoveUnusedCode.traverseNode(RemoveUnusedCode.java:448)
	at com.google.javascript.jscomp.RemoveUnusedCode.traverseDeclarationStatement(RemoveUnusedCode.java:1026)
	at com.google.javascript.jscomp.RemoveUnusedCode.traverseNode(RemoveUnusedCode.java:500)
	at com.google.javascript.jscomp.RemoveUnusedCode.traverseChildren(RemoveUnusedCode.java:1245)
	at com.google.javascript.jscomp.RemoveUnusedCode.traverseNode(RemoveUnusedCode.java:530)
	at com.google.javascript.jscomp.RemoveUnusedCode.traverseChildren(RemoveUnusedCode.java:1245)
	at com.google.javascript.jscomp.RemoveUnusedCode.traverseNode(RemoveUnusedCode.java:530)
	at com.google.javascript.jscomp.RemoveUnusedCode.access$1200(RemoveUnusedCode.java:82)
	at com.google.javascript.jscomp.RemoveUnusedCode$Continuation.apply(RemoveUnusedCode.java:1762)
	at com.google.javascript.jscomp.RemoveUnusedCode.traverseAndRemoveUnusedReferences(RemoveUnusedCode.java:363)
	at com.google.javascript.jscomp.RemoveUnusedCode.process(RemoveUnusedCode.java:338)
	at com.google.javascript.jscomp.PhaseOptimizer$NamedPass.process(PhaseOptimizer.java:267)
	at com.google.javascript.jscomp.PhaseOptimizer.process(PhaseOptimizer.java:186)
	at com.google.javascript.jscomp.Compiler.performTranspilationAndOptimizations(Compiler.java:2572)
	at com.google.javascript.jscomp.Compiler.lambda$stage2Passes$5(Compiler.java:936)
	at com.google.javascript.jscomp.CompilerExecutor.lambda$runInCompilerThread$0(CompilerExecutor.java:101)
	at java.util.concurrent.FutureTask.run(FutureTask.java:264)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
	at java.lang.Thread.run(Thread.java:829)
	at com.oracle.svm.core.thread.JavaThreads.threadStartRoutine(JavaThreads.java:597)
	at com.oracle.svm.core.posix.thread.PosixJavaThreads.pthreadStartRoutine(PosixJavaThreads.java:194)
```
</details>

kleisauke commented on src/parseTools.js:
```diff
@@ -37,10 +37,12 @@ function processMacros(text) {
 function preprocess(text, filenameHint) {
   if (EXPORT_ES6 && USE_ES6_IMPORT_META) {
     // `eval`, Terser and Closure don't support module syntax; to allow it,
```
Yes, eventually it would be nice to fix Terser and Closure, but it's a relatively large change.

For Terser we need to cherry-pick https://github.com/terser/terser/commit/24ce297dfc6b298251d031bc9a4f89c005a853fa on top of https://github.com/emscripten-core/terser/tree/emscripten_patches_v4.8.0 and Closure needs to run with some additional flags.
https://github.com/emscripten-core/emscripten/blob/234940f032e4353a4390af02bbc59a6a9f87afcf/tools/building.py#L603-L610

Though, I think(?), `eval()` still needs this replacement, since `import.meta` is used in `library_pthread.js` as well.
https://github.com/emscripten-core/emscripten/blob/234940f032e4353a4390af02bbc59a6a9f87afcf/src/modules.js#L215-L222
https://github.com/emscripten-core/emscripten/blob/234940f032e4353a4390af02bbc59a6a9f87afcf/emscripten.py#L207-L210

The above changes are not included in this PR, see comment https://github.com/emscripten-core/emscripten/pull/17915#issuecomment-1311871913.

RReverser replied:
> Though, I think(?), `eval()` still needs this replacement, since `import.meta` is used in `library_pthread.js` as well.

Pretty sure this was the reason I used this hack in the first place btw - it was just easier to support the variety of tooling + `eval` this way instead of doing something custom for each of them separately.

kleisauke replied:
> Pretty sure this was the reason I used this hack in the first place btw - it was just easier to support the variety of tooling + `eval` this way instead of doing something custom for each of them separately.

👍, I realised that too late, after doing all those Terser and Closure changes. 😅

Anyway, the Closure change, that optimizes the wrapper function when linking with `-sMODULARIZE`, might still be useful. I'll see if I can split that up after this PR lands.

kleisauke commented:
I just pinged the folks at #11792 to verify if this works for all uses-cases. Should I also post something at https://groups.google.com/g/emscripten-discuss?

RReverser commented:
I really think it would be worth to merge this so that folks can try it w/o fiddling with emsdk. Considering that Node.js+ESM build was unusable before this PR anyway, it should _at least_ not be a breaking change, and will be an easy way to collect feedback if any on the new mode.

RReverser commented:
It's a bit hard to tell with so many comments above (I tried to resolve a few conversations, but gave up), but I think there are no more blockers here?

kleisauke commented:
> I really think it would be worth to merge this so that folks can try it w/o fiddling with emsdk. Considering that Node.js+ESM build was unusable before this PR anyway, it should _at least_ not be a breaking change, and will be an easy way to collect feedback if any on the new mode.

You're right, this should be safe. I'll amend my above comment to mention using the `tot` target only, after this PR is landed and auto-rolled.

> It's a bit hard to tell with so many comments above (I tried to resolve a few conversations, but gave up), but I think there are no more blockers here?

As far as I know, all comments have been resolved/addressed. I'm a bit reluctant to land this PR by myself or press the "Resolve Converstation"-button as an external contributor.

sbc100 commented:
@kleisauke and @RReverser if you both agree this is step in the right direction please go ahead and land.  We can always make further changes as followups.

RReverser commented:
Done, thanks.

kleisauke commented:
This PR failed to auto-roll due to a failed test case on the macOS runner, see:
https://logs.chromium.org/logs/emscripten-releases/buildbucket/cr-buildbucket/8797224671288030801/+/u/Emscripten_testsuite__other_/stdout

Looks like Node/V8 failed on an assertion(?), curious. I'll investigate.

kleisauke commented:
... I opened PR https://github.com/emscripten-core/emscripten/pull/18230 for that.

RReverser commented:
Yet another reason to finally update Node.js in emsdk to supported versions...

lgarron commented:
> Since this PR has landed, it can be tested by installing the "tip-of-tree" build:

Fabulous, thank you for the instructions! 😻
I'm able to confirm that this allows `https://github.com/cubing/twsearch/` to run directly in `node`. 🤩

lgarron commented:
I do have a small nit, though, which is that the imports are still not prefixed with `node:`, e.g. `node:module`. This is necessary to distinguish the import from an `npm` package name. Would that be most appropriate to file a new issue for?

lgarron commented:
> I do have a small nit, though, which is that the imports are still not prefixed with `node:`, e.g. `node:module`. This is necessary to distinguish the import from an `npm` package name. Would that be most appropriate to file a new issue for?

I've created https://github.com/emscripten-core/emscripten/pull/17915#issuecomment-1320648665 for this.

RReverser commented:
I think `node:` prefix is a relatively recent addition in Node 18+, and not something supported in Node 14 that Emscripten ships with. Besides, I'm not sure what the bundler & other tooling support is like.

I'd rather err on the side of compatibility with more tooling for now.

paulcdejean commented:
Hello this is a neat new feature but I don't see any documentation of it anywhere.
